src/sources/syslog.rs

```diff
@@ -142,6 +143,106 @@ impl TcpSource for SyslogTcpSource {
     }
 }
 
+/// Decodes according to `Octet Counting` in https://tools.ietf.org/html/rfc6587
+#[derive(Clone, Debug)]
+struct SyslogDecoder {
+    other: LinesCodec,
+}
+
+impl SyslogDecoder {
+    fn new(max_length: usize) -> Self {
+        Self {
+            other: LinesCodec::new_with_max_length(max_length),
+        }
+    }
+
+    fn octet_decode(&self, src: &mut BytesMut) -> Result<Option<String>, io::Error> {
+        // Encoding scheme:
+        //
+        // len ' ' data
+        // |    |  | len number of bytes that contain syslog message
+        // |    |
+        // |    | Separating whitespace
+        // |
+        // | ASCII decimal number of unknown length
+
+        if let Some((i, _)) = src.iter().enumerate().find(|&(_, &b)| b == b' ') {
+            let len: usize = std::str::from_utf8(&src[..i])
+                .map_err(|_| ())
+                .and_then(|num| num.parse().map_err(|_| ()))
+                .map_err(|_| {
+                    io::Error::new(
+                        io::ErrorKind::InvalidData,
+                        "Unable to decode message len as number",
+                    )
+                })?;
+
+            let from = i + 1;
+            let to = from + len;
+
+            if let Some(msg) = src.get(from..to) {
+                let s = std::str::from_utf8(msg)
+                    .map_err(|_| {
+                        io::Error::new(
+                            io::ErrorKind::InvalidData,
+                            "Unable to decode message as UTF8",
+                        )
+                    })?
+                    .to_string();
+                src.advance(to);
+                Ok(Some(s))
+            } else {
+                Ok(None)
+            }
+        } else if src.len() < self.other.max_length() {
+            Ok(None)
+        } else {
+            // This is certainly mallformed, and there is no recovering from this.
+            Err(io::Error::new(
+                io::ErrorKind::Other,
+                "frame length limit exceeded",
+            ))
+        }
+    }
+
+    /// None if this is not octet counting encoded
+    fn checked_decode(&self, src: &mut BytesMut) -> Option<Result<Option<String>, io::Error>> {
+        if let Some(&first_byte) = src.get(0) {
+            if 49 <= first_byte && first_byte <= 57 {
+                // First character is non zero number so we can assume that
+                // octet count framing is used.
+                trace!("Octet counting encoded event detected.");
+                return Some(self.octet_decode(src));
```

Comment:
This makes me a little bit nervous. We try not to make too many assumptions about the syslog "format", and I'm not sure if there's any variant we accept that start with a number.

I also don't love the fact that this could theoretically change modes on a per-read basis. It's very clever and the UX of auto-detecting the format is great, I just want to be sure that it won't lead to confusing or incorrect behavior. I don't think the alternative of adding a configuration option is that onerous.

Reply:
The [`rfc6587`](https://tools.ietf.org/html/rfc6587) isn't quite a standard, but more of an observation, so yes some caution is needed. 

In code we always first match on `<` which is used as a start in both `rfc5424` and `rfc3164` so at the moment we don't have other formats that start with a number. 

> change modes on a per-read basis

We can detect format on first message and assume it on the rest of the stream. It has been noted in `rfc6587` that changing format on read per read basis is not a good practice, but has been observed in legacy systems.

It is not a strong guarantee, but again the `rfc6587` noted that:
> It can be assumed that octet-counting framing is used if a syslog frame starts with a digit.

Having configuration option would mean that user:
1. Has to know which formatting they will have.
2. If they have sources with different formatting they will need to have two `syslog` sources, one for each formatting.

Going by `rfc` only, I would say it's fine to auto detect.

Reply:
That makes sense to me. If the RFC says we can do it, that's good enough :smile: